### PR TITLE
Updates hydrogen.config and preps alerts, buttons, links, and standard cards for dark mode

### PR DESCRIPTION
**Pull request by substrae**
## ✨ Summary

Contact @substrae with design questions

This pull request preps the `hydrogen.config.json` file for dark mode, and begins the updates required for iterative component upgrades. @tristan-orourke might need to take a peek at this initial PR, as he mentioned potentially putting dark mode behind a feature flag as these iterative PRs are released and merged. I might also need @esizer to review the theme hooks/provider/story integrations as a result of these changes and the feature flag.

## 🔮 Details

This pull request addresses the following:
- Color updates in the Hydrogen config
- New story format for the components outlined below that showcases all 4 mode and theme combinations at once, allowing for more efficient design updates and accessibility testing (see screenshot included below)
- Style updates to the `alert` component
- Refactor and updates to the `button` component
- Refactor and updates to the `link` component
- Refactor and updates to the `downloadCsv` component
- Prop and style updates to the `card` component
- Prop and style updates to the `cardFlat` component
- Prop and style updates to the `cardLink` component

The prop updates mentioned above bring the card variants closer inline with each other so that they all support things like subtitles, the same colors, etc.

<details>
  <summary>📸 Story example</summary>
  
</details>

### CODE REVIEW

File: packages/ui/src/components/Styleguide/Styleguide.tsx [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            </div>
+          </div>
+          <div data-h2="dark">
+            <div
+              data-h2-background="base(background.dark)"
+              data-h2-padding="base(x1)"
+              data-h2-border="base(1px solid background.darkest)"
+              data-h2-color="base(background.darkest)"
+              data-h2-radius="base(5px)"
+            >
```
Comment (esizer) [inline packages/ui/src/components/Styleguide/Styleguide.tsx:184]: Looks like we also have a colour contrast issue here. Currently it is `3.63:1`

Screenshot 2023-09-20 110429

**Review by esizer — CHANGES_REQUESTED** [review]
This is looking really great. Love the new style guide story!

I just found some colour contrast issues that should be easy to solve.

File: packages/ui/src/components/Styleguide/Styleguide.tsx (hunk 1 above)
Comment (substrae) [reply packages/ui/src/components/Styleguide/Styleguide.tsx:184]: This fella will be updated as a part of the message wells later, but I'll note it down to be sure.

File: packages/ui/src/components/CardLink/CardLink.stories.tsx [hunk 2]
```diff
@@ -36,37 +37,106 @@ export default {
...
+        data-h2-grid-template-columns="base(1fr) base:children[>div>div](repeat(7, minmax(0, 1fr)))"
+        data-h2-height="base:children[>div>div](100%)"
+        data-h2-padding="base:children[>div>div](x2)"
+        data-h2-gap="base:children[>div>div](x1)"
+        data-h2-background-color="base:children[>div>div](background)"
+      >
+        <div data-h2="light">{Cards}</div>
+        <div data-h2="dark">{Cards}</div>
+        <div data-h2="iap light">{Cards}</div>
+        <div data-h2="iap dark">{Cards}</div>
```
[the remark this answers is not in the document]
Comment (substrae) [reply packages/ui/src/components/CardLink/CardLink.stories.tsx:136]: I'm still working on buttons, so these should all get fixed with the next push.

File: packages/ui/src/components/Button/Button.tsx [hunk 3]
```diff
@@ -29,17 +36,27 @@ const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
...
+      <button
+        ref={ref}
+        disabled={disabled}
+        {...getBaseStyle({ mode })}
+        {...getBackgroundColor({ mode, color, disabled })}
+        {...getBorderColor({ mode, color, disabled })}
+        {...getDisplay({ mode, block })}
+        {...getFontColor({ mode, color, disabled })}
+        {...getFontWeight({ mode })}
+        {...getShadow({ mode, disabled })}
```
Comment (tristan-orourke) [inline packages/ui/src/components/Button/Button.tsx:50]: I'm not against moving these functions into their own files, but I think it was nice to have a single function defined somewhere which combines them.

Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx:50]: Separating them was mostly for my ADHD brain - I couldn't focus on the existing file you guys were using because there was so much happening. Happy to add a consolidation function if you think it optimizes things.

File: packages/theme/src/components/ThemeProvider.tsx [hunk 4]
```diff
@@ -89,16 +89,18 @@ const ThemeProvider = ({
   );
 
   React.useEffect(() => {
-    const hydrogen = document.querySelectorAll(themeSelector || "[data-h2]");
+    const hydrogen = document.querySelectorAll(
+      themeSelector || "body[data-h2]",
+    );
     let themeString: string | undefined = "";
     // TO DO: Add mode back once dark mode is done
```
[the remark this answers is not in the document]
Comment (substrae) [reply packages/theme/src/components/ThemeProvider.tsx:96]: Ideally @esizer can review that aspect of this PR (or even submit a separate one) that re-enables dark mode generally, and puts it behind a feature flag

File: packages/ui/src/components/Button/Button.tsx (hunk 3 above)
Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx]: Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7946/commits/b65ccf6fa9b5e37028d9f7d74d79d8c7763fbc56

File: packages/theme/src/components/ThemeProvider.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (esizer) [reply packages/theme/src/components/ThemeProvider.tsx:96]: Okay, so two things I noticed here:

1. The `[data-h2]` is actually on the `html` element so the query selector doesn't return anything meaning the theme will never update (should change it to `html[data-h2]` so we can use the IAP theme)
2. The comment was poorly placed by me. It refers to the overwriting of the the string on line 99 which prevents it from ever setting `mode` and only ever sets the `theme` portion

https://github.com/GCTC-NTGC/gc-digital-talent/blob/a86f69e21e8542326b675f5e270622cf040c51ed/packages/theme/src/components/ThemeProvider.tsx#L97-L99

**Review by esizer — CHANGES_REQUESTED** [review]
I already let you know but putting it here so we don't forget 😅 

Other than the CTA buttons on smaller screens this is looking great 👌 

Screenshot 2023-09-28 094841

File: packages/ui/src/components/Button/Button.stories.tsx [hunk 5]
```diff
@@ -54,6 +54,7 @@ const Template: Story = () => {
         data-h2-display="base(grid)"
         data-h2-grid-template-columns="base(1fr 1fr 1fr 1fr)"
         data-h2-text-align="base(center)"
+        data-h2-color="base(red)"
```
Comment (esizer) [inline packages/ui/src/components/Button/Button.stories.tsx:57]: While I test this, should we remove these?

**Review by esizer — APPROVED** [review]
Sweet! 🍭 

---

I think this is ready to go once we merge in the admin theme changes! 🎉 . I couldn't find anything else but it may be good to get someone else to very quickly run through maybe while you get the admin theme ready 🤔

File: packages/ui/src/components/Button/Button.stories.tsx (hunk 5 above)
Comment (substrae) [reply packages/ui/src/components/Button/Button.stories.tsx:57]: Honestly, I kind of like keeping them there just as a safety... Things will be real obvious if they stop working for whatever reason.

File: packages/ui/src/components/Link/Link.tsx
Comment (petertgiles) [inline packages/ui/src/components/Link/Link.tsx:1]: Some our underlines are missing.  This is intentional?
image
image

Comment (petertgiles) [inline packages/ui/src/components/Link/Link.tsx:1]: Are these link changes (text alignment, icon size) intentional?
image

File: packages/storybook-helpers/decorators/ThemeDecorator.tsx
Comment (petertgiles) [inline packages/storybook-helpers/decorators/ThemeDecorator.tsx:1]: Seems like the IAP theme stopped working.
image
image

File: packages/ui/src/components/Button/Button.tsx
Comment (petertgiles) [inline packages/ui/src/components/Button/Button.tsx:1]: Seems like the conditional display mode on the IAP buttons stopped working.
image

Comment (petertgiles) [inline packages/ui/src/components/Button/Button.tsx:1]: This one is a bit awkward.
image

Comment (petertgiles) [inline packages/ui/src/components/Button/Button.tsx:1]: Was this change intentional?
image

Comment (petertgiles) [inline packages/ui/src/components/Button/Button.tsx:1]: Is this change intentional?
image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Very cool!  :sunglasses: I just finished looking through the Chromatic diffs and flagged a few things that I wasn't sure about.

**Comment by petertgiles** [comment]
I really like the new story formats!  I'm hoping we can factor out the boilerplate from the story files into something reusable but having all the variants in a single story like that is super nice.

File: packages/theme/src/components/ThemeProvider.tsx [hunk 6]
```diff
@@ -89,16 +89,18 @@ const ThemeProvider = ({
...
-    const hydrogen = document.querySelectorAll(themeSelector || "[data-h2]");
+    const hydrogen = document.querySelectorAll(
+      themeSelector || "html[data-h2]",
+    );
     let themeString: string | undefined = "";
-    // TO DO: Add mode back once dark mode is done
     if (mode && key) {
-      // themeString = `${key} ${mode}`;
+      themeString = `${key} ${mode}`;
+      // TO DO: Remove this line to re-enable dark mode
```
Comment (petertgiles) [inline packages/theme/src/components/ThemeProvider.tsx:98]: I tried removing L99 but I didn't see any difference.  Are there any stories that look different with theme mode changing?

File: packages/ui/src/components/Alert/Alert.stories.tsx [hunk 7]
```diff
@@ -25,32 +25,50 @@ const TemplateAlert: ComponentStory<typeof Alert.Root> = ({
...
-          </Alert.Root>
-        </React.Fragment>
-      ))}
+    <div>
+      <div
+        data-h2-display="base(grid) base:children[>div>div](grid)"
+        data-h2-grid-template-columns="base(1fr) base:children[>div>div](repeat(2, minmax(0, 1fr)))"
+        data-h2-padding="base:children[>div>div](x2)"
+        data-h2-gap="base:children[>div>div](x1)"
+        data-h2-background-color="base:children[>div>div](background)"
```
Comment (petertgiles) [inline packages/ui/src/components/Alert/Alert.stories.tsx:57]: I really like the new format of story showing all the different variants together!  I don't love all this boilerplate in teh story but hopefully we can eventually figure out how to factor it out to something reusable.

File: packages/ui/src/components/Button/Button.tsx
Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx:1]: @esizer Any chance you can take a look at whether the changes I made to the theme provider caused this?

File: packages/theme/src/components/ThemeProvider.tsx (hunk 6 above)
Comment (esizer) [reply packages/theme/src/components/ThemeProvider.tsx:98]: This should be just for the main application. I do not believe this will affect anything in storybook.

File: packages/ui/src/components/Button/Button.tsx
Comment (esizer) [reply packages/ui/src/components/Button/Button.tsx:1]: Looks like it was missing the width when displayed as block. I always thought `display: flex;` made it fill the horizontal area but I guess not 😓 

Fixed with [fix block button width](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7946/commits/6c0bf749910037463fea31e66c180642c9fabbe5)

File: packages/ui/src/components/Link/Link.tsx
Comment (substrae) [reply packages/ui/src/components/Link/Link.tsx:1]: Intentional for the accordions, not for the links. I'll take a peek .

Comment (substrae) [reply packages/ui/src/components/Link/Link.tsx:1]: So these are generally formatted weird, even on `main`. I'll take a look at fixing them. The icon size thing is on purpose though.

File: packages/ui/src/components/Alert/Alert.stories.tsx (hunk 7 above)
Comment (substrae) [reply packages/ui/src/components/Alert/Alert.stories.tsx:57]: I agree, boilerplate is kinda gross, but at the same time, it's not 100% the same across stories. Some components require more vertical/horizontal space, so the grid changes based on their individual needs. Open to ways of trying to standardize this though!

File: packages/ui/src/components/Link/Link.tsx
Comment (substrae) [reply packages/ui/src/components/Link/Link.tsx:1]: @petertgiles Do you mind reviewing this bit again now that I've pushed up a fix?

File: packages/ui/src/components/Button/Button.tsx
Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx:1]: This is actually going to be more work to fix than anticipated. I'm going to fix it properly though, but adding a 5th button mode called `placeholder` so that we can use this elsewhere if need be.

Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx:1]: @petertgiles This might have been fixed with Eric's block-style changes. Mind double checking?

Comment (substrae) [reply packages/ui/src/components/Button/Button.tsx:1]: @petertgiles Okay this should be good to go now :) And we have shiny new placeholder buttons in all colors that can leverage the existing props

**Comment by petertgiles** [comment]
@substrae could you please resolve the merge conflict so we can run another Chromatic diff?

File: packages/ui/src/components/Link/Link.tsx
Comment (petertgiles) [reply packages/ui/src/components/Link/Link.tsx:1]: Oh yeah, this looks way better!

File: packages/storybook-helpers/decorators/ThemeDecorator.tsx
Comment (petertgiles) [reply packages/storybook-helpers/decorators/ThemeDecorator.tsx:1]: I'm still seeing this in Chromatic.

File: packages/ui/src/hooks/useCommonButtonLinkStyles.ts
Comment (petertgiles) [inline packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: It looks like the "scroll to" links got left out of the bolding changes.
image

File: packages/ui/src/utils/button/getButtonBaseStyle.ts
Comment (petertgiles) [inline packages/ui/src/utils/button/getButtonBaseStyle.ts:1]: Do you want these ToC buttons to be left-aligned?
image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Sorry, a few more things to check.  :cry:

File: packages/ui/src/hooks/useCommonButtonLinkStyles.ts
Comment (substrae) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: So this might be a distinction between `text` and `inline` style links. `text` should be used when placing a link within content, whereas `inline` should be used in places where you could also use a standard button, but want to draw less attention to it.

File: packages/storybook-helpers/decorators/ThemeDecorator.tsx
Comment (substrae) [reply packages/storybook-helpers/decorators/ThemeDecorator.tsx:1]: I'm very confused by this. Hm..

File: packages/ui/src/hooks/useCommonButtonLinkStyles.ts
Comment (petertgiles) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: It looks closer now, but not exactly the same.
image
The regular links are weight 400 and the last one is 300.

File: packages/ui/src/utils/button/getButtonBaseStyle.ts
Comment (petertgiles) [reply packages/ui/src/utils/button/getButtonBaseStyle.ts:1]: Hmm, I'm still seeing this.
image

File: packages/ui/src/hooks/useCommonButtonLinkStyles.ts
Comment (substrae) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: @petertgiles Thanks for being patient with these - I noticed the hero element was actually really out of whack with the Figma designs, so I've gone and updated:
- whitespace
- subtitle font size
- subtitle font weight (everything is 400 now)

File: packages/ui/src/utils/button/getButtonBaseStyle.ts
Comment (substrae) [reply packages/ui/src/utils/button/getButtonBaseStyle.ts:1]: @petertgiles My bad on this one - I applied the centering to buttons but not links. Should be good now.

File: packages/theme/src/components/ThemeProvider.tsx [hunk 8]
```diff
@@ -89,16 +89,18 @@ const ThemeProvider = ({
   );
 
   React.useEffect(() => {
-    const hydrogen = document.querySelectorAll(themeSelector || "[data-h2]");
+    const hydrogen = document.querySelectorAll(
+      themeSelector || "body[data-h2]",
```
Comment (esizer) [inline packages/theme/src/components/ThemeProvider.tsx:93]: This needs to be `html` to work with the app since that is where we have the root `data-h2` attribute. I think this change was made for storybook, correct? Maybe we could try targeting both? 🤔 

```suggestion
      themeSelector || "body[data-h2], html[data-h2]",
```

Comment (substrae) [reply packages/theme/src/components/ThemeProvider.tsx]: Gave this a test and it looks like it works to me. Wanna pull it down and give it a look for me?

Comment (petertgiles) [reply packages/theme/src/components/ThemeProvider.tsx]: Works for me.  :+1:

File: packages/ui/src/hooks/useCommonButtonLinkStyles.ts
Comment (petertgiles) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: OK, it's definitely bigger now.  :laughing:   I see that the links are the same size but I'm noticing that the hover is broken in the hero and in the cards.  This might have happened earlier since I was doing most of my reviews in Chromatic where it wasn't interactive.  Here's a side-by-side video with prod.

https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/6f4ce5a2-6d8a-480f-b5fe-633fada9c6e5

Comment (substrae) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: @petertgiles Turns out this was a bit more widespread than I anticipated. I hope my latest changes haven't overstepped, but I've:
- Added the button/link styles to `ScrollToLink`
- Updated the one-off `StatusItem` links in the `HeroCard`

Comment (petertgiles) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: I'm seeing the hover style in the status items not but I'm not seeing it yet on the hero subtitle or the hero card titles.  Is it working for you?

https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/5e3ea769-135c-4ecb-b124-d5f5137f85db

Comment (substrae) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: Weird, yeah, mine is working. I'll triple check that I'm not misremembering and get back to you.

Comment (substrae) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: @petertgiles Yeah, super weird... I see it
image

Comment (petertgiles) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: Sorry about that.  OK, I'll do a fresh rebuild today.

Comment (petertgiles) [reply packages/ui/src/hooks/useCommonButtonLinkStyles.ts:1]: Yep, you're right.  After a fresh build it's looking good.  :+1:

**Review by petertgiles — APPROVED** [review]
Wow, it's a big one!  Thanks for sticking with it.  I don't see anything else that jumps out to me.  :medal_sports:
